common.go

```diff
@@ -0,0 +1,51 @@
+package set
+
+// Common is the interface that all sets implement
+type Common[T any] interface {
+	// Slice returns a slice of all elements in the set
+	Slice() []T
+	// Insert inserts an element into the set
+	// if the element already exists, it will return false
+	Insert(T) bool
+	// InsertSlice inserts all elements from the slice into the set
+	InsertSlice([]T) bool
+	// Size returns the number of elements in the set
+	Size() int
+	// ForEach  will call the callback function for each element in the set.
+	// If the callback returns false, the iteration will stop.
+	// Note: iteration order depends on the underlying implementation;
+	ForEach(call func(T) bool)
+}
+
+// InsertSliceFunc inserts all elements from the slice into the set
+func InsertSliceFunc[T, E any](s Common[T], items []E, f func(element E) T) {
+	for _, item := range items {
+		s.Insert(f(item))
+	}
+}
+
+// TransformUnion transforms the set A into another set B
+func TransformUnion[T, E any](a Common[T], b Common[E], transform func(T) E) {
+	a.ForEach(func(item T) bool {
+		_ = b.Insert(transform(item))
+		return true
+	})
+}
+
+// Transform transforms the set into another set
+// newSet is a  function that return a new set
+func Transform[T, E any](a Common[T], transform func(T) E, newSet func() Common[E]) Common[E] {
+	b := newSet()
+	TransformUnion(a, b, transform)
+	return b
```

Comment:
Ahhh I didn't think of the `newSet` problem. Sorry to keep you going back and forth @zonewave, but can you remove `Transform` for now. I want to think if there's a cleaner way we can do this, without having to pass in a function to construct a new set. There's probably some generics magic + type switching we can do here, but I need time to think about it.

Reply:
actually, I don't think it looks elegant.I have removed the last submission